homeassistant/components/cookidoo/coordinator.py

```diff
@@ -0,0 +1,71 @@
+"""DataUpdateCoordinator for the Cookidoo integration."""
+
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+
+from cookidoo_api import (
+    Cookidoo,
+    CookidooAuthException,
+    CookidooException,
+    CookidooItem,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_EMAIL
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, TODO_ADDITIONAL_ITEMS, TODO_INGREDIENTS
+
+_LOGGER = logging.getLogger(__name__)
+
+type CookidooConfigEntry = ConfigEntry[CookidooDataUpdateCoordinator]
+
+
+class CookidooDataUpdateCoordinator(
+    DataUpdateCoordinator[dict[str, list[CookidooItem]]]
+):
+    """A Cookidoo Data Update Coordinator."""
+
+    config_entry: CookidooConfigEntry
+
+    def __init__(self, hass: HomeAssistant, cookidoo: Cookidoo) -> None:
+        """Initialize the Cookidoo data coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=90),
+        )
+        self.cookidoo = cookidoo
+
+    async def _async_update_data(self) -> dict[str, list[CookidooItem]]:
+        try:
+            ingredients = await self.cookidoo.get_ingredients()
+            additional_items = await self.cookidoo.get_additional_items()
+        except CookidooAuthException as e:
+            try:
+                await self.cookidoo.refresh_token()
+            except CookidooAuthException as exc:
+                raise ConfigEntryAuthFailed(
+                    translation_domain=DOMAIN,
+                    translation_key="setup_authentication_exception",
+                    translation_placeholders={
+                        CONF_EMAIL: self.config_entry.data[CONF_EMAIL]
+                    },
+                ) from exc
+            raise UpdateFailed(
+                "Authentication failed but re-authentication was successful, trying again later"
+            ) from e
+        except CookidooException as e:
+            raise UpdateFailed(
+                "Unable to connect and retrieve data from cookidoo"
+            ) from e
+
+        return {
+            TODO_INGREDIENTS: ingredients,
+            TODO_ADDITIONAL_ITEMS: additional_items,
+        }
```

Comment:
did you consider the suggestion from tr4nt0r to make this a dataclass (or a TypedDict at least)? Would make it easier to know what to expect here